core/scale/variant.hpp

```diff
@@ -41,60 +41,53 @@ namespace kagome::scale::variant {
 
     template <class... T>
     struct VariantEncoder {
-      using Variant = std::variant<T...>;
+      using Variant = boost::variant<T...>;
       using Result = outcome::result<void>;
-      const Variant &v;
-      Buffer &out;
-      Result &res;
+      const Variant &v_;
+      Buffer &out_;
+      Result &res_;
 
       VariantEncoder(const Variant &v, Buffer &buf, Result &res)
-          : v{v}, out{buf}, res{res} {}
+          : v_{v}, out_{buf}, res_{res} {}
 
       template <class H>
       void apply(uint8_t index) {
         // if type matches alternative in variant then encode
-        if (std::holds_alternative<H>(v)) {
+        kagome::visit_in_place(v_, [this, index](const H &h) {
           // first byte means type index
-          out.putUint8(index);
-          // get alternative ptr
-          auto *alternative = std::get_if<H>(&v);
-          if (nullptr == alternative) {
-            res = EncodeError::WRONG_ALTERNATIVE;
-            return;
-          }
-          // encode the value using custom type encoder
-          auto &&encode_result = TypeEncoder<H>{}.encode(*alternative, out);
+          out_.putUint8(index);
+          auto &&encode_result = TypeEncoder<H>{}.encode(h, out_);
           if (!encode_result) {
-            res = encode_result.error();
+            res_ = encode_result.error();
           }
-        }
+        }, [](const auto &) {});
       }
     };
 
     template <class... T>
     struct VariantDecoder {
```

Comment:
VariantDecoder should be eventually reworked as TypeDecoder

Reply:
added TODOs